BDF font support

projectitis opened the pull request:
Created new Resource type hxd.res.BDFFont that extends Resource.  Use it to load/parse a BDF font files to h2d.Font instance.

Example usage:

`var f = hxd.Res.My_BDF_font_file.toFont();`
`var tf = new h2d.Text(f);`

A BDF font file (format created by Adobe, possibly in the 1980s) is a self-contained text file that contains packed bit data with optional smoothing/alpha and does not have a separate PNG bitmap.

The BDFFont resource parses the BDF file and creates it's own glyph tile. It uses a basic but pretty effective packing algorithm to generate (almost) the smallest tile it can to save space.

FontForge is an example program that is able to generate BDF files. The benefit of BDF files is that the entire font is self-contained. A negative of the BDF format is that it does not have support for kerning.

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,437 @@
+     * @return Int      The final line number after processing header
+     */
+     @:access(h2d.Font)
+     private function parseFontHeader( lines : Array<String>, linenum : Int ) : Int{
```
Code style does not use explicit `private function`, just `private`, same goes for `private static inline var`

projectitis replied:
Sorry, I'm not sure what you mean. Do you mean that `private function parseFontHeader(...)` should read `private parseFontHeader(...)`? I have not seen this elsewhere in the source. or do you mean I shouldn't be using the `private` keyword?

Yanrishatum replied:
Yes, if access not specified it defaults to `private`, and Heaps does not use explicit declaration for private variables and methods.

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,437 @@
+     * @param s     The string
+     * @return Int  The resulting integer
+     */
+    private function extractInt( s : String ) : Int{
```
`inline function`

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,437 @@
+ * parsing a BDF font file.
+ */
+ class BDFFontChar{
+    public var code : Int;
```
Code style uses tabs, not spaces.

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,436 @@
+#end
+
+import hxd.BitmapData;
+import haxe.Int32;
```
Is there a particular reason you need to explicitly specify it's an Int32? If no, better to just use `Int`, which is a core type and not an abstract with overhead.

projectitis replied:
It holds an ARGB value and I wanted to be sure Int was wide enough. If Int is always at least 32 bit then I'm happy to change to Int?

Yanrishatum replied:
Unless Haxe (and Heaps) somehow starts to support 16-bit target with standard int being less than 32-bit in size - it is guaranteed to be at least of that size.

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,436 @@
+ * Intermediate representation of a glyph. Only used while
+ * parsing a BDF font file.
+ */
+ class BDFFontChar{
```
CodeStyle: All opening brackets should contain space before them (`expr {`)

projectitis replied:
Do you also prefer I remove brackets from `case` or also ensure there is a space before them?
`case 'ENCODING': { ... }`

Yanrishatum replied:
Huh, didn't notice brackets on cases. But yes, better remove them, as they are completely unnecessary.

projectitis replied:
Yes, I just like them for readability. Will remove them. Will always keep them in my own code :)

projectitis commented:
Well, I figured the font gets parsed once on load and that is all (not, for example, every render loop) so sacrificing performance for clarity was the initial goal.  Once this is part of the library, please feel free to optimise it - or feel free to suggest optimisations and I can add :)

projectitis commented:
Please excuse my ignorance. No offence intended (this is my first contribution to another repo), but is there something else I need to do before this is merged?

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+				switch (prop) {
+					case 'ENCODING':
+						// XXX: Support encoding ranges. Not sure if they are common? Hacen't come across this yet
+						if (encoding!=-1) throw ('Encoding ranges not supported, at line '+linenum);
+						encoding = extractInt(args[0]);
+						if (encoding<1) throw('ENCODING '+encoding+' not supported, at line '+linenum);
+					case 'DWIDTH': // Device width
+						stride = extractInt(args[0]);
+						if (stride<0) throw('DWIDTH is negative, at line '+linenum);
+						// XXX: This could be a warning and we could ignore the vertical step. Font might render weird?
+						if (extractInt(args[1])!=0) throw('A non-0 DWIDTH is not supported (maybe vertical character set?), at line '+linenum);
```
```suggestion
						if ( encoding != -1 ) throw 'Encoding ranges not supported, at line $linenum';
						encoding = extractInt(args[0]);
						if ( encoding < 1 ) throw 'ENCODING $encoding not supported, at line $linenum';
					case 'DWIDTH': // Device width
						stride = extractInt(args[0]);
						if ( stride < 0 ) throw 'DWIDTH is negative, at line $linenum';
						// XXX: This could be a warning and we could ignore the vertical step. Font might render weird?
						if ( extractInt(args[1]) != 0 ) throw 'A non-0 DWIDTH is not supported (maybe vertical character set?), at line $linenum';
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+						bbxHeight = extractInt(args[1]);
+						bbxXOffset = extractInt(args[2]);
+						bbxYOffset = extractInt(args[3]);
+						if (bbxWidth < 0) throw('BBX width is negative, line '+linenum);
+						if (bbxHeight < 0) throw('BBX height is negative, line '+linenum);
+					case 'BITMAP':
+						if (encoding<0) throw('missing ENCODING, line '+linenum);
+						if (stride<0) throw('missing DWIDTH, line '+linenum);
+						if (!bbxFound) throw('missing BBX, line '+linenum);
```
```suggestion
						if ( bbxWidth < 0 ) throw 'BBX width is negative, line $linenum';
						if ( bbxHeight < 0 ) throw 'BBX height is negative, line $linenum';
					case 'BITMAP':
						if ( encoding < 0 ) throw 'missing ENCODING, line $linenum';
						if ( stride < 0 ) throw 'missing DWIDTH, line $linenum';
						if ( !bbxFound ) throw 'missing BBX, line $linenum';
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			} // header
+
+			// Secondly, extract the bitmap data (will be processed later)
+			else{
```
```suggestion
			else {
```

Yanrishatum replied:
ditto

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			// Secondly, extract the bitmap data (will be processed later)
+			else{
+				// Extract data as long as we are still expecting some
+				if ((expectedLines > 0) && (expectedBytesPerLine > 0)) {
+					for (i in 0...expectedBytesPerLine) {
```
```suggestion
				if ( expectedLines > 0 && expectedBytesPerLine > 0 ) {
					for ( i in 0...expectedBytesPerLine ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+					expectedLines--;
+				}
+				// Otherwise we have finished reading bitmap data
+				else{
```
```suggestion
				else {
```

Yanrishatum replied:
ditto

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+				// Otherwise we have finished reading bitmap data
+				else{
+					// Sanity check
+					if (prop != 'ENDCHAR') throw('ENDCHAR expected, line '+linenum);
```
```suggestion
					if ( prop != 'ENDCHAR' ) throw 'ENDCHAR expected, line $linenum';
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+		var volume : Int = 0;
+		for (d in glyphData) volume += (d.width*d.height);
+		var bitmapWidth : Int = Math.ceil(Math.sqrt(volume*(1+BitmapPad)));
+		if (bitmapWidth>BitmapMaxWidth) throw('The font bitmap is too big: '+bitmapWidth+'x'+bitmapWidth+' (max '+BitmapMaxWidth+'x'+BitmapMaxWidth+')');
```
```suggestion
		if ( bitmapWidth > BitmapMaxWidth ) throw 'The font bitmap is too big: ${bitmapWidth}x$bitmapWidth (max ${BitmapMaxWidth}x$BitmapMaxWidth)';
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+		// Calculate total volume, and from that an approx width and height if packing with 80%
+		// efficiency (i.e. add a 10% buffer). This is from trial and error :)
+		var volume : Int = 0;
+		for (d in glyphData) volume += (d.width*d.height);
```
```suggestion
		for ( d in glyphData ) volume += (d.width * d.height);
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+
+		// Calculate values for extracting pixel data
+		var bppMask : Int = 0x80;
+		switch (bitsPerPixel) {
```
```suggestion
		switch ( bitsPerPixel ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			case 4: bppMask = 0xF0;
+			case 8: bppMask = 0xFF;
+		}
+		var pixPerByte : Int = Math.floor(8/bitsPerPixel);
+		var bppScale : Float = 255/((1 << bitsPerPixel)-1);
```
```suggestion
		var pixPerByte : Int = Math.floor(8 / bitsPerPixel);
		var bppScale : Float = 255 / ((1 << bitsPerPixel) - 1);
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+		// efficiency (i.e. add a 10% buffer). This is from trial and error :)
+		var volume : Int = 0;
+		for (d in glyphData) volume += (d.width*d.height);
+		var bitmapWidth : Int = Math.ceil(Math.sqrt(volume*(1+BitmapPad)));
```
```suggestion
		var bitmapWidth : Int = Math.ceil(Math.sqrt(volume * (1 + BitmapPad)));
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+		var x : Int = 0;
+		var y : Int = 0;
+		var found : Bool = false;
+		for (d in glyphData) {
```
```suggestion
		for ( d in glyphData ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			found = false;
+
+			// Wrap x if glyph will not fit in width
+			if ((x+d.width) > bitmapWidth) x = 0;
```
```suggestion
			if ( x + d.width > bitmapWidth ) x = 0;
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			if ((x+d.width) > bitmapWidth) x = 0;
+
+			// Find nearest space big enough for glyph, left to right, top to bottom
+			while (x<=(bitmapWidth-d.width)) {
+				y = 0;
+				while (y<=(bitmapWidth-d.height)) {
+					// If top-left pixel is clear...
+					if (bitmapData.getPixel(x,y)==ClearColor) {
+						found = true;
+						// Check first row and first column are clear to ensure space is clear
+						for (xx in x...(x+d.width)) {
+							if (bitmapData.getPixel(xx,y)!=ClearColor) {
```
```suggestion
			while ( x <= bitmapWidth - d.width ) {
				y = 0;
				while ( y <= bitmapWidth - d.height ) {
					// If top-left pixel is clear...
					if ( bitmapData.getPixel(x,y) == ClearColor ) {
						found = true;
						// Check first row and first column are clear to ensure space is clear
						for ( xx in x...(x+d.width) ) {
							if ( bitmapData.getPixel(xx,y) != ClearColor ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+								break;
+							}
+						}
+						if (found) {
+							for (yy in y...(y+d.height)) {
+								if (bitmapData.getPixel(x,yy)!=ClearColor) {
```
```suggestion
						if ( found ) {
							for ( yy in y...(y+d.height) ) {
								if ( bitmapData.getPixel(x,yy) != ClearColor ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+								}
+							}
+						}
+						if (found) break;
```
```suggestion
						if ( found ) break;
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+					}
+					y++;
+				}
+				if (found) break;
```
```suggestion
				if ( found ) break;
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			}
+
+			// XXX: At this point it would be really good to see the bitmap (so far)
+			if (!found) throw('Glyphs are overflowing the bitmap. Help!');
```
```suggestion
			if ( !found ) throw 'Glyphs are overflowing the bitmap. Help!';
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			// Now have space that starts at x,y.
+			// Draw the glyph to the bitmap and save position
+			d.x = x; d.y = y;
+			for (yy in y...(y+d.height)) {
+				pixLeftInByte = 0;
+				for (xx in x...(x+d.width)) {
+					// Grab a new byte
+					if (pixLeftInByte==0) {
```
```suggestion
			for ( yy in y...(y+d.height) ) {
				pixLeftInByte = 0;
				for ( xx in x...(x+d.width) ) {
					// Grab a new byte
					if ( pixLeftInByte == 0 ) {
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+						pixBits = d.bits.shift();
+					}
+					// Grab a pixel alpha
+					pixAlpha = (pixBits & bppMask) >> (8-bitsPerPixel);
```
```suggestion
					pixAlpha = (pixBits & bppMask) >> (8 - bitsPerPixel);
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+		font.tile = h2d.Tile.fromBitmap(bitmapData);
+
+		// Generate glyphs
+		for (d in glyphData) {
+			// In BDF, y-offset is offset from baseline. In FNT it appears to be offset from top
+			var t = font.tile.sub(d.x, d.y, d.width, d.height, d.xoffset, ascent-(d.height+d.yoffset) );
```
```suggestion
		for ( d in glyphData ) {
			// In BDF, y-offset is offset from baseline. In FNT it appears to be offset from top
			var t = font.tile.sub(d.x, d.y, d.width, d.height, d.xoffset, ascent - d.height + d.yoffset );
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+			font.glyphs.set(" ".code, new h2d.Font.FontChar(font.tile.sub(0, 0, 0, 0), font.size>>1));
+
+		// Set line height, or estimate it
+		if ((ascent>=0) && (descent>=0))
+			font.lineHeight = ascent + descent;
+		else if ((fbbHeight>=0))
+			font.lineHeight = fbbHeight;
+		else{
+			var a = font.glyphs.get("E".code);
+			if ( a == null )
+				a = font.glyphs.get("A".code);
+			if (a==null)
```
```suggestion
		if ( ascent >= 0 && descent >= 0 )
			font.lineHeight = ascent + descent;
		else if ( fbbHeight >= 0 )
			font.lineHeight = fbbHeight;
		else {
			var a = font.glyphs.get("E".code);
			if ( a == null )
				a = font.glyphs.get("A".code);
			if ( a == null )
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+	inline function extractStr( p : Array<String> ) : String {
+		if (StringTools.startsWith(p[0],'"')) {
+			var pj : String = StringTools.trim(p.join(' '));
+			return pj.substring(1,pj.length-1);
```
```suggestion
			return pj.substring(1, pj.length - 1);
```

Yanrishatum commented on hxd/res/BDFFont.hx:
```diff
@@ -0,0 +1,415 @@
+	 * @param s     The string
+	 * @return Int  The resulting integer
+	 */
+	inline function extractInt( s : String ) : Int{
```
```suggestion
	inline function extractInt( s : String ) : Int {
```

projectitis commented:
ok, thank you. Will go through entire file taking particular care of formatting. Good to know that is considered important in a PR - how would I normally find out what is required when I contribute to the project? Is there a contributor help page?

projectitis commented:
Thanks for your patience!

I have reformatted the code as far as I can tell to your style suggestions, and it is much more readable. Please note that the original library is all over the place in terms of style and doesn't provide a good example of how contributions should be formatted. For example, much of my code was based on files like `hxd.fmt.bfnt.Fontparser.hx`, which is very inconsistent itself.  I took a look at other files at random, and they all are the same.  Even so, I do appreciate that you are trying to ensure all future contributions meet style guidelines.

If it was possible to have a short wiki article on style requirements for contributions, that would be great!

projectitis commented:
I squashed commits and rebased to master head [following this guide](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request), then forced a push of the branch. I hope this was the correct way to do it? If not, let me know.

ncannasse commented:
Tell me if ready for merge @Yanrishatum ;)

Yanrishatum commented:
@ncannasse It's ready to merge since I clicked that "Approve" button ;)
And for RFC - it's not ready for simple reason - tech is not here, apparently. I tried.

ncannasse commented:
Merged ! Thanks for the contribution !
